### PR TITLE
Add pipeline support

### PR DESCRIPTION
**Pull request by wbagdon**
First shot at adding pipeline support to Memory Map plugin.

Main problem seems to be with adding graphs to the syntax generator.
It seems to be describable and has a config.jelly, not sure why it isn't being generated properly.

### CODE REVIEW

**Comment by buep** [comment]
Thanks for the effort - we will see if we can guide or help you.

@MadsNielsen or @praqma-thi  when you have time, will you look at this PR ?

**Comment by buep** [comment]
@praqma-thi or @MadsNielsen - did you have time to look at it? 

@wbagdon do you need help to any specifics regarding your PR? Else @MadsNielsen or @praqma-thi will just go over your changes regarding accepting them.

**Comment by wbagdon** [comment]
Still not entirely sure how to get the `MemoryMapGraphConfiguration` object to get generated properly in the snippet generator, so right now it is still not working

**Comment by MadsNielsen** [comment]
Hey @wbagdon, thanks for the contribution! :) 

It's going to be a challenge, since the recommended way to work with pipeline is to use configuration objects which all use simple types in their databound constructors. 
MemoryMap plugin uses a _slightly_ different approach where each object is a rather complex amalgamation of different options.

We'll need to provide some new constructors which take simplified arguments in order to make it easier to configure the job. So basically, we'll have to replace the arrays of configuration objects in the constructors with simple strings, integers, etc. That alone should make it less painful to get the snippet generator working.

Just for fun, we tried configuring the step manually. We got bored after writing half a novel, and it doesn't even configure half the step.
```
node("master") {
  step([$class: "net.praqma.jenkins.memorymap.MemoryMapRecorder", chosenParsers: [[$class: "net.praqma.jenkins.memorymap.parser.gcc.GccMemoryMapParser", parserUniqueName: "uniqueName", mapFile: "map.map", configurationFile: "linker.ld", wordSize: 8, bytesOnGraph: true, graphConfiguration: [], pattern: ".*"   ]], showBytesOnGraph: true, wordSize: 8, scale: "words", graphConfiguration: []])  
}
```

We only managed a brief look after a hectic week, we'll dig a little deeper soon and get back to you next week.

**Comment by wbagdon** [comment]
I think I can cut down on the "novel writing" by implementing some defaults with `@DataBoundSetter` extensions.  This is the first plugin I'm dealing with, so any guidance is welcome.

**Comment by wbagdon** [comment]
Implementing the databound setters seems to have fixed the snippet generator and I am getting output in the console related to the configuration.

```
Memory Map Plugin version 2.1.3-SNAPSHOT (private-5ed35d9c-bbagdon)
Printing configuration

{default=[0] RAM [origin = 0x1C00, length = 0x2800, used = 00000671, unused = 0000218f, endAddress = N/A]
}
```

I'm not seeing the graph on either the individual job, or the pipeline page.
Any ideas on how to get that running?

**Comment by praqma-thi** [comment]
Hey @wbagdon, thanks for getting the snippet generator up and running, that was a big help.
We took a look at your problem with the graph rendering.

The results to render are added to the build as a list of `MemoryMapProjectAction`. 
The rendering is done on the job/build page through the action's `index.jelly`.
After looking around a bit, I think `MemoryMapRecorder` is what's putting a stick in the wheels.

Traditionally, `MemoryMapRecorder`'s `getProjectAction` gets called to fetch the actions we'd like to render, but considering that signature takes an `AbstractProject` and not a `Run`, it's likely we're missing something here. There's no overload available that takes a `Run`, either.

There's mention of a `TransientActionFactory<Job>` in the dev guide, which sounds pretty relevant. It's overload returns a list of actions, which should be our `MemoryMapProjectAction`s. Sadly, `MemoryMapProjectAction` makes heavy use of `AbstractProject`, it might be a pain to set it up to use the `Run` instead.

**TL;DR**
Replacing `MemoryMapRecorder`'s `getProjectAction` with a new `TransientActionFactory<Job>` that returns the relevant actions is probably a good place to start.

I'm not that familiar with implementing Pipeline support either, but I hope that gives you a good starting point. Let us know if you have any questions at all.

**Comment by wbagdon** [comment]
OK I think this should give me a good starting point

**Comment by MadsNielsen** [comment]
@wbagdon Also check this out: https://jenkins.io/blog/2016/05/25/update-plugin-for-pipeline/ 

There is a section "Build Actions vs. Project Actions" which describes what needs to be implemented in order to show your project action.

**Comment by wbagdon** [comment]
I've been using that along with these two other sources for the conversion:
https://github.com/jenkinsci/pipeline-plugin/blob/master/DEVGUIDE.md
https://www.youtube.com/watch?v=azyv183Ua6U

**Comment by wbagdon** [comment]
I've ported this over and it seems to work with both Freestyle and Pipeline jobs on the project page.
Known issues right now:

- Not seeing the graph on the individual builds
- Projects had to be made from scratch to be able to see the graphs

**Comment by MadsNielsen** [comment]
Hey @wbagdon thanks for the contribution! We'll review your changes internally and test them out. 

We'll probably bump major version to 3.0, since you took the opportunity to clean up some old backwards compatibility code in the `readResolve()` methods. Spring cleaning was due, thanks! 👍

**Comment by wbagdon** [comment]
There's a few other deprecated methods still left if you want me to clean those up too

**Comment by praqma-thi** [comment]
You definitely won't hear me complaining. :wink: 

We took the plugin for a spin last night, the snippet generator is pumping everything out correctly and the plugin steps get executed with all the right configuration, so everything's great there.

I didn't get any further as I found out some of the tests are failing on me, only tested it on my local environment, though. IIRC, the integration tests that are failing are excluded from the default maven test target, so you probably weren't notified if you accidentally broke something.

I'll take a look at them Monday and see what's up.

**Comment by wbagdon** [comment]
@praqma-thi it had to do with my implementation of `RoleChecker` in the move to 1.580.1
I reverted the change, not sure what the ramifications of this are so you may want to look at this.

**Comment by wbagdon** [comment]
There are 13 remaining findbug errors (5 of which have to deal with `Jenkins.getInstance()` now having a `@CheckForNull` annotation.

We could bump the required Jenkins version to the next LTS (1.596.1) to take advantage of `getActiveInstance()` that was added in 1.590 to get around the findbug errors thrown by `getInstance()`
Or add `<failOnError>false</failOnError>` to the findbugs POM

But I'm not sure if how you want to fix (or ignore) this, so I'll leave it up to you.

**Comment by praqma-thi** [comment]
We'll just bump the core version and use `getActiveInstance()` as you mentioned. After @MadsNielsen and I merge in your changes and fix the remaining FindBugs issues, we'll push out a new release.
Thanks for your efforts, Bill. :+1:
